### PR TITLE
DSpace: archive format option

### PR DESCRIPTION
**Pull request by sevein**
Redmine: https://projects.artefactual.com/issues/10808

7z used to be the default. This commit makes ZIP the new default and updates
the DSpace model so the user can choose between the two.

### CODE REVIEW

File: storage_service/locations/migrations/0012_dspace_archive_format.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='dspace',
+            name='archive_format',
+            field=models.CharField(default=b'ZIP', max_length=3, verbose_name=b'Archive format', choices=[(b'ZIP', b'ZIP'), (b'7Z', b'7z')]),
+        ),
+
+        migrations.RunPython(data_migration),
```
Comment (sevein) [inline storage_service/locations/migrations/0012_dspace_archive_format.py]: I realize this is probably not necessary, right?

Comment (sevein) [reply storage_service/locations/migrations/0012_dspace_archive_format.py]: Should I use `preserve_default=True`?

File: storage_service/locations/models/dspace.py [hunk 2]
```diff
@@ -121,6 +128,58 @@ def _get_metadata(self, input_path, aip_uuid):
...
+
+    def _archive_zip(self, src, dst):
+        """ Return the command that creates the ZIP archive file. """
+        if not dst.endswith('.zip'):
+            dst += '.zip'
+
+        return (dst, [
+            'zip',
+            '-r',
+            dst,  # Destination (without extension)
```
Comment (Hwesta) [inline storage_service/locations/models/dspace.py]: Should `dst` have the extension here? The comment says no but the current behaviour is yes.

File: storage_service/locations/models/dspace.py [hunk 3]
```diff
@@ -121,6 +128,58 @@ def _get_metadata(self, input_path, aip_uuid):
         os.remove(mets_path)
         return kwargs
 
+    def _archive(self, src, dst):
+        """
+        Run the archiver in order to create the archive file based on the
+        preferred format. `dst` is the path of the archive file. The file
+        extension must not be included, instead this function will return
+        the final destination with the extension on it according to the
+        archive format preferred.
```
Comment (Hwesta) [inline storage_service/locations/models/dspace.py]: Tip: [PEP257](https://www.python.org/dev/peps/pep-0257/) for docstrings suggests having a commit-message-like docstring: single line summary, optionally followed by a multi-line explanation.

File: storage_service/locations/models/dspace.py [hunk 4]
```diff
@@ -152,43 +211,11 @@ def _split_package(self, input_path):
...
-            '-mmt=on',  # Multithreaded
-            metadata_zip,  # Destination
-            metadata_dir,  # Source
-        ]
-        try:
-            subprocess.check_call(command)
-        except subprocess.CalledProcessError:
-            LOGGER.error('Could not compress %s', metadata_dir)
-            raise
+        metadata_zip = self._archive(metadata_dir, os.path.join(output_dir, 'metadata'))
```
Comment (Hwesta) [inline storage_service/locations/models/dspace.py:231]: Nice deduplication of code here.

File: storage_service/locations/tests/test_dspace.py [hunk 5]
```diff
@@ -74,6 +76,15 @@ def test_split_package(self):
         split_paths = self.dspace_object._split_package(path)
         # Verify
         assert len(split_paths) == 2
+        assert os.path.join(FIXTURES_DIR, 'objects.zip') in split_paths
```
Comment (Hwesta) [inline storage_service/locations/tests/test_dspace.py:79]: Maybe rename this to `test_split_package_zip` ?

File: storage_service/locations/migrations/0012_dspace_archive_format.py (hunk 1 above)
Comment (Hwesta) [reply storage_service/locations/migrations/0012_dspace_archive_format.py]: I suspect it's not necessary, but you can verify that.

**Review by Hwesta — APPROVED** [review]
This is more configurable than I was expecting for this change, but looks like it works.  

We're looking at implementing `move_to_storage_service` (redmine 10862) for DSpace, which will require dealing with what happens if the package format changes between storing and retrieving.

File: storage_service/locations/models/dspace.py (hunk 2 above)
Comment (sevein) [reply storage_service/locations/models/dspace.py]: ZIP doesn't really care but the comment was definitely wrong, will fix.

**Comment by sevein** [comment]
I've addressed your comments, thank you!

**Review by Hwesta — APPROVED** [review]

**Comment by sevein** [comment]
Merged. Accidentally used cherry-pick: a3bc8dba07b80c41af5794fa75c5fd0313ce60cf.

**Pull request closed without merging**